feat: add C implementation for `stats/base/dists/normal/mgf`

0PrashantYadav0 opened the pull request:
Resolves #3778

## Description

> What is the purpose of this pull request?

This pull request:

-   adds C implementation for @stdlib/stats/base/dists/normal/mgf along with relevant tests, docs, examples and benchmarks.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #3778

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Neerajpathak07 committed:
refactor: update `stats/base/dnanmeanpn` native addon from C++ to C

PR-URL: https://github.com/stdlib-js/stdlib/pull/4110
Reviewed-by: Athan Reines <kgryte@gmail.com>

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: passed
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: passed
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

0PrashantYadav0 committed:
Merge branch 'develop' of https://github.com/0PrashantYadav0/stdlib into develop

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

0PrashantYadav0 committed:
Merge branch 'develop' of https://github.com/0PrashantYadav0/stdlib into develop

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

0PrashantYadav0 committed:
Merge branch 'stdlib-js:develop' into develop

0PrashantYadav0 committed:
feat: add C implementation for normal mgf

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

stdlib-bot committed:
chore: update copyright years

0PrashantYadav0 commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into normal/mgf

0PrashantYadav0 committed:
fix: changes in benchmarks

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/normal/mgf/README.md:
```diff
@@ -156,6 +156,98 @@ for ( i = 0; i < 10; i++ ) {
+The function accepts the following arguments:
+
+-   **t**: `[in] double` input value.
+-   **mu**: `[in] double` mean of the distribution
+-   **sigma**: `[in] double` standard deviation of the distribution
```
```suggestion
-   **mu**: `[in] double` mean.
-   **sigma**: `[in] double` standard deviation.
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/normal/mgf/README.md:
```diff
@@ -156,6 +156,98 @@ for ( i = 0; i < 10; i++ ) {
+}
+
+int main( void ) {
+    double t;
+    double mu;
+    double sigma;
+    double y;
+    int i;
```
```suggestion
    double sigma;
    double mu;
    double t;
    double y;
    int i;
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/normal/mgf/README.md:
```diff
@@ -156,6 +156,98 @@ for ( i = 0; i < 10; i++ ) {
+    for ( i = 0; i < 10; i++ ) {
+        t = random_uniform( 0.0, 1.0 );
+        mu = random_uniform( -50.0, 50.0 );
+        sigma = random_uniform( 0.0, 20.0 ) + STDLIB_CONSTANT_FLOAT64_EPS;
```
```suggestion
        sigma = random_uniform( STDLIB_CONSTANT_FLOAT64_EPS, 20.0 );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/normal/mgf/benchmark/benchmark.js:
```diff
@@ -32,17 +33,25 @@ var mgf = require( './../lib' );
+	for ( i = 0; i < len; i++ ) {
+		t[ i ] = uniform( 0.0, 1.0 );
+		mu[ i ] = uniform( -50.0, 50.0 );
+		sigma[ i ] = uniform( 0.0, 20.0 ) + EPS;
```
```suggestion
		sigma[ i ] = uniform( EPS, 20.0 );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/normal/mgf/benchmark/benchmark.native.js:
```diff
@@ -0,0 +1,74 @@
+	for ( i = 0; i < len; i++ ) {
+		t[ i ] = uniform( 0.0, 1.0 );
+		mu[ i ] = uniform( -50.0, 50.0 );
+		sigma[ i ] = uniform( 0.0, 20.0 ) + EPS;
```
```suggestion
		sigma[ i ] = uniform( EPS, 20.0 );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/normal/mgf/benchmark/c/benchmark.c:
```diff
@@ -0,0 +1,144 @@
+* @return elapsed time in seconds
+*/
+static double benchmark( void ) {
+	double elapsed;
+	double t[ LEN ];
+	double mu[ LEN ];
+	double sigma[ LEN ];
+	double y;
+	double tc;
+	int i;
```
You mind ordering these by length, please?

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/normal/mgf/benchmark/c/benchmark.c:
```diff
@@ -0,0 +1,144 @@
+	for ( i = 0; i < LEN; i++ ) {
+		t[ i ] = random_uniform( -100.0, 100.0 );
+		mu[ i ] = random_uniform( -50.0, 50.0 );
+		sigma[ i ] = random_uniform( 0.0, 20.0 ) + STDLIB_CONSTANT_FLOAT64_EPS;
```
```suggestion
		sigma[ i ] = random_uniform( STDLIB_CONSTANT_FLOAT64_EPS, 20.0 );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/normal/mgf/examples/c/example.c:
```diff
@@ -0,0 +1,44 @@
+	for ( i = 0; i < 10; i++ ) {
+		t = random_uniform( 0.0, 1.0 );
+		mu = random_uniform( -50.0, 50.0 );
+		sigma = random_uniform( 0.0, 20.0 ) + STDLIB_CONSTANT_FLOAT64_EPS;
```
```suggestion
		sigma = random_uniform( STDLIB_CONSTANT_FLOAT64_EPS, 20.0 );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/normal/mgf/examples/c/example.c:
```diff
@@ -0,0 +1,44 @@
+#include "stdlib/constants/float64/eps.h"
+#include <stdlib.h>
+#include <stdio.h>
+#include <time.h>
```
Why is this header included?

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/normal/mgf/README.md:
```diff
@@ -156,6 +156,98 @@ for ( i = 0; i < 10; i++ ) {
+#include "stdlib/constants/float64/eps.h"
+#include <stdlib.h>
+#include <stdio.h>
+#include <time.h>
```
Why is this header included?

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/normal/mgf/src/main.c:
```diff
@@ -0,0 +1,46 @@
+* Evaluates the moment-generating function (MGF) for a normal distribution with mean `mu` and standard deviation `sigma` at a value `t`.
+*
+* @param t      input value
+* @param mu     mean of the distribution
+* @param sigma  standard deviation of the distribution
```
```suggestion
* @param mu     mean
* @param sigma  standard deviation
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/normal/mgf/manifest.json:
```diff
@@ -0,0 +1,85 @@
+        "@stdlib/math/base/napi/ternary",
+        "@stdlib/math/base/assert/is-nan",
+        "@stdlib/math/base/special/pow",
+        "@stdlib/math/base/special/exp",
+        "@stdlib/constants/float64/eps"
```
```suggestion
        "@stdlib/math/base/special/exp"
```

EPS is mot used in `main.c` or `addon.c`.

Planeshifter reviewed (changes requested):
Thanks, @0PrashantYadav0 for your PR. Left a bunch of comments that should be resolved.

0PrashantYadav0 committed:
chore: updated according to code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: failed
---

0PrashantYadav0 committed:
fix: add missing include for stdio.h in normal mgf README

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

0PrashantYadav0 committed:
fix: add missing include for stdio.h in example.c

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: passed
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: passed
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/normal/mgf/README.md:
````diff
@@ -156,6 +156,97 @@ for ( i = 0; i < 10; i++ ) {
+}
+```
+
 </section>
````
```suggestion
</section>

<!-- /.examples -->

</section>

<!-- /.c -->
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/normal/mgf/examples/c/example.c:
```diff
@@ -0,0 +1,43 @@
+}
+
+int main( void ) {
+	double t;
+	double mu;
+	double sigma;
+	double y;
+	int i;
```
```suggestion
	double sigma;
	double mu;
	double t;
	double y;
	int i;
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/normal/mgf/benchmark/benchmark.native.js:
```diff
@@ -0,0 +1,74 @@
+
+// MAIN //
+
+bench( pkg, opts, function benchmark( b ) {
```
```suggestion
bench( pkg+'::native', opts, function benchmark( b ) {
```

Planeshifter committed:
chore: apply suggestions from code review

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter reviewed (approved):
Thanks for all your efforts, @0PrashantYadav0!
